src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -557,3 +523,46 @@ exprCheckMatchPatterns d ctx e@(EMatch _ x cs) = do
     assert (consTreeEmpty ct) (pos x) "Non-exhaustive match patterns"
 
 exprCheckMatchPatterns _ _   _               = return ()
+
+-- Automatically insert string conversion functions in the Concat
+-- operator:  '"x:" ++ x', where 'x' is of type int becomes 
+-- '"x:" ++ int_2string(x)'.
+progInjectStringConversions :: (MonadError String me) => DatalogProgram -> me DatalogProgram
+progInjectStringConversions d = progExprMapCtxM d (exprInjectStringConversions d)
+
+exprInjectStringConversions :: (MonadError String me) => DatalogProgram -> ECtx -> ENode -> me Expr
+exprInjectStringConversions d ctx e@(EBinOp p Concat l r) | (te == tString) && (tr /= tString) = do
+    -- find string conversion function
+    fname <- case tr of
+                  TBool{}     -> return bUILTIN_2STRING_FUNC
+                  TInt{}      -> return bUILTIN_2STRING_FUNC
+                  TString{}   -> return bUILTIN_2STRING_FUNC
+                  TBit{}      -> return bUILTIN_2STRING_FUNC
+                  TUser{..}   -> return $ mk2string_func typeName
+                  TOpaque{..} -> return $ mk2string_func typeName
+                  TTuple{}    -> err (pos r) "Automatic string conversion for tuples is not supported"
+                  TVar{..}    -> err (pos r) $ 
+                                     "Cannot automatically convert " ++ show r ++ 
+                                     " of variable type " ++ tvarName ++ " to string"
+                  TStruct{}   -> error "unexpected TStruct in exprInjectStringConversions"
+    f <- case lookupFunc d fname of
+              Nothing  -> err (pos r) $ "Cannot find declaration of function " ++ fname ++ 
+                                        " needed to convert expression " ++ show r ++ " to string"
+              Just fun -> return fun
+    let arg0 = funcArgs f !! 0
+    -- validate its signature
+    assert (isString d $ funcType f) (pos f) 
```

Comment:
I don't know what assert does in Haskell, but in general you should not report errors that are caused by user programs using assertions.

Reply:
this is our assert function that will generate a nice compilation error

Reply:
Shall I call it `check`?

Reply:
That's a good name.